File: tests/interop/interop_test.go
// Cross-implementation interop checks: a third-party MASQUE client drives
// CONNECT-UDP sessions through a running volto server.
//
// Every other test in this repository has volto on both ends — the in-process
// HTTP/3 client in `tests/common/mod.rs` is built from the same pinned `h3`
// revision as the server, so the two agree by construction even where both are
// wrong. This suite removes that symmetry: the client is Go's
// github.com/quic-go/masque-go on top of quic-go, an implementation that shares
// no code, no QUIC stack and no reading of the RFCs with volto.
//
// It is driven by the `interop` job of .github/workflows/ci.yml, which starts a
// real volto process and passes its address, SNI and certificate in the
// environment. Nothing here starts or configures the server.
package interop

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"net"
	"net/http"
	"os"
	"testing"
	"time"

	masque "github.com/quic-go/masque-go"
	"github.com/quic-go/quic-go"
	"github.com/quic-go/quic-go/http3"
	"github.com/yosida95/uritemplate/v3"
)

const (
	// Address the volto server under test listens on, as host:port.
	envAddr = "VOLTO_ADDR"
	// Name to validate the server certificate against. Defaults to "localhost".
	envSNI = "VOLTO_SNI"
	// PEM file holding the server's self-signed certificate, which is trusted
	// as a root here rather than skipping verification: an interop test that
	// does not verify the certificate would not notice a TLS regression.
	envCert = "VOLTO_CERT"
	// Credentials sent as HTTP Basic on every CONNECT request. Both empty means
	// the server under test has authentication disabled.
	envUser     = "VOLTO_USER"
	envPassword = "VOLTO_PASSWORD"
)

// Bound on any single exchange. Everything here is loopback, so this only ever
// fires when something is actually broken.
const exchangeTimeout = 10 * time.Second

// A port volto's default policy refuses (SMTP), used to drive a refusal.
const deniedPort = 25

// dialProxy opens one QUIC connection to the server under test.
//
// The connection is returned as a masque.ClientConn, on which any number of
// CONNECT-UDP sessions can be opened — that multiplexing is the whole point of
// MASQUE, and the case this suite cares most about.
func dialProxy(t *testing.T, ctx context.Context) *masque.ClientConn {
	t.Helper()

	addr := os.Getenv(envAddr)
	if addr == "" {
		t.Fatalf("%s is not set: this suite tests a server started by CI, it does not start one", envAddr)
	}

	sni := os.Getenv(envSNI)
	if sni == "" {
		sni = "localhost"
	}

	certPath := os.Getenv(envCert)
	if certPath == "" {
		t.Fatalf("%s is not set: the server certificate has to be trusted explicitly", envCert)
	}
	pem, err := os.ReadFile(certPath)
	if err != nil {
		t.Fatalf("reading %s = %q: %v", envCert, certPath, err)
	}
	roots := x509.NewCertPool()
	if !roots.AppendCertsFromPEM(pem) {
		t.Fatalf("%s = %q contains no certificate", envCert, certPath)
	}

	conn, err := quic.DialAddr(ctx, addr, &tls.Config{
		RootCAs:    roots,
		ServerName: sni,
		NextProtos: []string{http3.NextProtoH3},
	}, &quic.Config{
		EnableDatagrams: true,
		// Large enough that a tunnelled QUIC handshake would still fit; also
		// what masque-go's own Transport defaults to.
		InitialPacketSize: 1350,
	})
	if err != nil {
		t.Fatalf("QUIC handshake with %s failed: %v", addr, err)
	}
	t.Cleanup(func() { _ = conn.CloseWithError(0, "") })

	client, err := (&masque.Transport{}).NewClientConn(conn)
	if err != nil {
		t.Fatalf("HTTP/3 client connection: %v", err)
	}
	return client
}

// proxyTemplate builds the RFC 9298 §2 default URI template for the server.
//
// Surge has no way to configure a template either, so this is the same URI
// shape the production client uses.
func proxyTemplate(t *testing.T) *uritemplate.Template {
	t.Helper()

	sni := os.Getenv(envSNI)
	if sni == "" {
		sni = "localhost"
	}
	_, port, err := net.SplitHostPort(os.Getenv(envAddr))
	if err != nil {
		t.Fatalf("%s = %q is not host:port: %v", envAddr, os.Getenv(envAddr), err)
	}

	raw := fmt.Sprintf("https://%s/.well-known/masque/udp/{target_host}/{target_port}/", net.JoinHostPort(sni, port))
	tmpl, err := uritemplate.New(raw)
	if err != nil {
		t.Fatalf("parsing URI template %q: %v", raw, err)
	}
	return tmpl
}

// newRequest builds a CONNECT-UDP request for target, carrying credentials when
// the environment supplies them.
func newRequest(t *testing.T, ctx context.Context, tmpl *uritemplate.Template, target string) *masque.Request {
	t.Helper()

	req, err := masque.NewRequest(ctx, tmpl, target)
	if err != nil {
		t.Fatalf("building a CONNECT-UDP request for %s: %v", target, err)
	}
	if user := os.Getenv(envUser); user != "" {
		// The header Surge was observed to use. Sending it from an unrelated
		// client is what proves the server does not depend on anything
		// Surge-specific about how it is framed.
		req.Header().Set("Proxy-Authorization", basicCredentials(user, os.Getenv(envPassword)))
	}
	return req
}

// basicCredentials renders an HTTP Basic field value (RFC 7617).
func basicCredentials(user, password string) string {
	const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"

	plain := []byte(user + ":" + password)
	token := make([]byte, 0, 4*((len(plain)+2)/3))
	for i := 0; i < len(plain); i += 3 {
		chunk := plain[i:min(i+3, len(plain))]
		var bits uint32
		for j := 0; j < 3; j++ {
			bits <<= 8
			if j < len(chunk) {
				bits |= uint32(chunk[j])
			}
		}
		for j := 0; j < len(chunk)+1; j++ {
			token = append(token, alphabet[(bits>>(18-6*j))&0x3f])
		}
		for j := len(chunk) + 1; j < 4; j++ {
			token = append(token, '=')
		}
	}
	return "Basic " + string(token)
}

// startEchoTarget runs a UDP server that echoes each packet back with tag
// prepended, and returns its address.
//
// The tag is what turns "a packet came back" into "the packet came back from
// the target this session was opened for": with several sessions in flight on
// one QUIC connection, a Quarter Stream ID mix-up shows up as a reply carrying
// somebody else's tag rather than as a lost packet.
func startEchoTarget(t *testing.T, tag byte) string {
	t.Helper()

	socket, err := net.ListenPacket("udp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("binding a UDP echo target: %v", err)
	}
	t.Cleanup(func() { _ = socket.Close() })

	go func() {
		buf := make([]byte, 65535)
		for {
			n, from, err := socket.ReadFrom(buf)
			if err != nil {
				return
			}
			reply := append([]byte{tag}, buf[:n]...)
			if _, err := socket.WriteTo(reply, from); err != nil {
				return
			}
		}
	}()

	return socket.LocalAddr().String()
}

// exchange sends payload through a session and returns what came back.
func exchange(t *testing.T, session *masque.Conn, payload []byte) []byte {
	t.Helper()

	if err := session.SetReadDeadline(time.Now().Add(exchangeTimeout)); err != nil {
		t.Fatalf("setting a read deadline: %v", err)
	}
	if _, err := session.WriteTo(payload, nil); err != nil {
		t.Fatalf("sending %d bytes to the target: %v", len(payload), err)
	}

	buf := make([]byte, 65535)
	n, _, err := session.ReadFrom(buf)
	if err != nil {
		t.Fatalf("reading the reply to %q: %v", payload, err)
	}
	return buf[:n]
}

// checkCapsuleResponse asserts the response fields RFC 9297 §3.4 requires of a
// message whose body is a capsule sequence.
func checkCapsuleResponse(t *testing.T, rsp *http.Response) {
	t.Helper()

	if got := rsp.Header.Get("Capsule-Protocol"); got != "?1" {
		t.Errorf("Capsule-Protocol = %q, want %q", got, "?1")
	}
	// RFC 9297 §3.4: a capsule-carrying message must not frame a body length.
	for _, name := range []string{"Content-Length", "Content-Type", "Transfer-Encoding"} {
		if got := rsp.Header.Get(name); got != "" {
			t.Errorf("%s = %q, want it absent", name, got)
		}
	}
}

// One session, several round trips: the baseline that the tunnel carries UDP
// payloads in both directions unmodified.
//
// More than one round trip on purpose. A Context ID or Quarter Stream ID that
// is only right for the first packet is a real failure mode (it is exactly what
// the h3-datagram release bug looked like), and a single-exchange test would
// pass straight through it.
func TestSingleSessionRoundTrips(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
	defer cancel()

	client := dialProxy(t, ctx)
	tmpl := proxyTemplate(t)
	target := startEchoTarget(t, 0xA0)

	session, rsp, err := client.Dial(newRequest(t, ctx, tmpl, target))
	if err != nil {
		t.Fatalf("CONNECT-UDP to %s: %v", target, err)
	}
	defer session.Close()

	if rsp.StatusCode != http.StatusOK {
		t.Fatalf("status = %d, want 200", rsp.StatusCode)
	}
	checkCapsuleResponse(t, rsp)

	for i := 0; i < 8; i++ {
		payload := []byte(fmt.Sprintf("volto-interop-packet-%d", i))
		reply := exchange(t, session, payload)

		want := append([]byte{0xA0}, payload...)
		if string(reply) != string(want) {
			t.Fatalf("round %d: reply = %q, want %q", i, reply, want)
		}
	}
}

// Three sessions on one QUIC connection, each with its own target.
//
// This is the regression baseline for the Quarter-Stream-ID class of bug: with
// every session multiplexed onto the same connection, misrouting a datagram
// does not lose it, it delivers it to the wrong session — which the per-target
// tag makes visible. All three sessions are written to before any of them is
// read, so the packets really are in flight at the same time.
func TestConcurrentSessionsDoNotCrossTalk(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
	defer cancel()

	client := dialProxy(t, ctx)
	tmpl := proxyTemplate(t)

	tags := []byte{0xB1, 0xB2, 0xB3}
	sessions := make([]*masque.Conn, len(tags))
	for i, tag := range tags {
		target := startEchoTarget(t, tag)
		session, rsp, err := client.Dial(newRequest(t, ctx, tmpl, target))
		if err != nil {
			t.Fatalf("session %d: CONNECT-UDP to %s: %v", i, target, err)
		}
		defer session.Close()

		if rsp.StatusCode != http.StatusOK {
			t.Fatalf("session %d: status = %d, want 200", i, rsp.StatusCode)
		}
		checkCapsuleResponse(t, rsp)
		sessions[i] = session
	}

	for round := 0; round < 4; round++ {
		payloads := make([][]byte, len(sessions))
		for i, session := range sessions {
			payloads[i] = []byte(fmt.Sprintf("session-%d-round-%d", i, round))
			if err := session.SetReadDeadline(time.Now().Add(exchangeTimeout)); err != nil {
				t.Fatalf("session %d: setting a read deadline: %v", i, err)
			}
			if _, err := session.WriteTo(payloads[i], nil); err != nil {
				t.Fatalf("session %d: sending: %v", i, err)
			}
		}

		for i, session := range sessions {
			buf := make([]byte, 65535)
			n, _, err := session.ReadFrom(buf)
			if err != nil {
				t.Fatalf("session %d round %d: reading the reply: %v", i, round, err)
			}

			want := append([]byte{tags[i]}, payloads[i]...)
			if string(buf[:n]) != string(want) {
				t.Fatalf("session %d round %d: reply = %q, want %q (a reply tagged for another session means datagrams are misrouted)",
					i, round, buf[:n], want)
			}
		}
	}
}

// A refusal has to be a well-formed HTTP response, not a dropped stream: a
// third-party client must be able to tell "denied" from "broken".
//
// The port is the one volto's default policy refuses, so this also proves the
// policy runs before anything is opened towards the target.
func TestDeniedTargetIsRefusedWithProxyStatus(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
	defer cancel()

	client := dialProxy(t, ctx)
	tmpl := proxyTemplate(t)
	target := net.JoinHostPort("127.0.0.1", fmt.Sprint(deniedPort))

	session, rsp, err := client.Dial(newRequest(t, ctx, tmpl, target))
	if err == nil {
		session.Close()
		t.Fatalf("CONNECT-UDP to %s succeeded, want a refusal", target)
	}
	if rsp == nil {
		t.Fatalf("CONNECT-UDP to %s failed without a response: %v", target, err)
	}
	if rsp.StatusCode != http.StatusForbidden {
		t.Fatalf("status = %d, want 403", rsp.StatusCode)
	}
	// RFC 9209: the reason is machine-readable, which is what makes a refusal
	// diagnosable from the client side.
	if got := rsp.Header.Get("Proxy-Status"); got != "volto; error=http_request_denied" {
		t.Errorf("Proxy-Status = %q, want %q", got, "volto; error=http_request_denied")
	}
}

// Credentials are checked before the request is routed, so a client that omits
// them is challenged rather than tunnelled.
//
// Skipped when the server under test has authentication disabled.
func TestMissingCredentialsAreChallenged(t *testing.T) {
	if os.Getenv(envUser) == "" {
		t.Skipf("%s is not set: the server under test has authentication disabled", envUser)
	}

	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
	defer cancel()

	client := dialProxy(t, ctx)
	tmpl := proxyTemplate(t)
	target := startEchoTarget(t, 0xC0)

	// Deliberately built without newRequest, so no credentials are attached.
	req, err := masque.NewRequest(ctx, tmpl, target)
	if err != nil {
		t.Fatalf("building a CONNECT-UDP request: %v", err)
	}

	session, rsp, err := client.Dial(req)
	if err == nil {
		session.Close()
		t.Fatal("an unauthenticated CONNECT-UDP succeeded, want 407")
	}
	if rsp == nil {
		t.Fatalf("CONNECT-UDP failed without a response: %v", err)
	}
	if rsp.StatusCode != http.StatusProxyAuthRequired {
		t.Fatalf("status = %d, want 407", rsp.StatusCode)
	}
	if got := rsp.Header.Get("Proxy-Authenticate"); got != `Basic realm="masque"` {
		t.Errorf("Proxy-Authenticate = %q, want %q", got, `Basic realm="masque"`)
	}
}

File: tests/interop/aioquic/interop_test.py
#!/usr/bin/env python3
"""Cross-implementation interop checks: aioquic drives volto.

The masque-go suite next door already removes the volto-on-both-ends symmetry
for CONNECT-UDP, but it leaves two holes this suite covers:

- masque-go speaks only CONNECT-UDP, so the plain CONNECT (TCP) tunnel has no
  cross-implementation check at all without this suite.
- masque-go owns the whole datagram encoding, so it cannot send a datagram
  with a non-zero Context ID and watch it being ignored; here the Context ID
  prefix is applied by hand, which is exactly what makes that test possible.

aioquic shares no code, no QUIC stack and no reading of the RFCs with either
volto or quic-go, so agreement here is a third independent vote.

Like the Go suite, this one is driven by the `interop` job of ci.yml: a real
volto process is started by CI and its address, SNI, certificate and
credentials arrive in the environment. Nothing here starts a server.
"""

import asyncio
import base64
import os
import sys

from aioquic.asyncio.client import connect
from aioquic.asyncio.protocol import QuicConnectionProtocol
from aioquic.buffer import Buffer, encode_uint_var
from aioquic.h3.connection import H3_ALPN, H3Connection
from aioquic.h3.events import DataReceived, DatagramReceived, HeadersReceived
from aioquic.quic.configuration import QuicConfiguration

# Bound on any single exchange. Everything is loopback; this fires only when
# something is actually broken.
EXCHANGE_TIMEOUT = 10.0

# A port volto's default policy refuses (SMTP), used to drive a refusal.
DENIED_PORT = 25


def env_addr():
    addr = os.environ.get("VOLTO_ADDR")
    if not addr:
        sys.exit("VOLTO_ADDR is not set: this suite tests a server started by CI")
    host, _, port = addr.rpartition(":")
    return host, int(port)


def env_sni():
    return os.environ.get("VOLTO_SNI", "localhost")


def basic_credentials():
    """The Proxy-Authorization value for the configured user, or None."""
    user = os.environ.get("VOLTO_USER", "")
    if not user:
        return None
    plain = f"{user}:{os.environ.get('VOLTO_PASSWORD', '')}".encode()
    return b"Basic " + base64.b64encode(plain)


class Refused(Exception):
    """A request drew a non-200 response; carries the response fields."""

    def __init__(self, status, headers):
        super().__init__(f"status {status}")
        self.status = status
        self.headers = headers


class ProxyClient(QuicConnectionProtocol):
    """One QUIC connection to volto, multiplexing any number of tunnels.

    The HTTP/3 layer is aioquic's own; this class only routes its per-stream
    events into queues the tests can await.
    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._h3 = H3Connection(self._quic, enable_webtransport=True)
        self._events = {}  # stream_id -> asyncio.Queue of H3 events

    def quic_event_received(self, event):
        for h3_event in self._h3.handle_event(event):
            stream_id = getattr(h3_event, "stream_id", None)
            if stream_id is not None:
                self._queue(stream_id).put_nowait(h3_event)

    def _queue(self, stream_id):
        return self._events.setdefault(stream_id, asyncio.Queue())

    async def _next_event(self, stream_id):
        return await asyncio.wait_for(
            self._queue(stream_id).get(), timeout=EXCHANGE_TIMEOUT
        )

    async def _request(self, headers):
        """Sends a CONNECT-shaped request and waits for the response fields.

        Returns the stream id on a 200; raises Refused otherwise.
        """
        stream_id = self._quic.get_next_available_stream_id()
        self._h3.send_headers(stream_id, headers, end_stream=False)
        self.transmit()

        event = await self._next_event(stream_id)
        assert isinstance(event, HeadersReceived), f"expected headers, got {event}"
        fields = dict(event.headers)
        status = int(fields[b":status"])
        if status != 200:
            raise Refused(status, fields)
        return stream_id, fields

    async def connect_tcp(self, authority):
        """Opens a plain CONNECT tunnel (RFC 9114 §4.4) to authority."""
        headers = [(b":method", b"CONNECT"), (b":authority", authority.encode())]
        credentials = basic_credentials()
        if credentials:
            headers.append((b"proxy-authorization", credentials))
        stream_id, _ = await self._request(headers)
        return stream_id

    async def connect_udp(self, host, port, with_credentials=True):
        """Opens a CONNECT-UDP session (RFC 9298) via the well-known template."""
        authority = f"{env_sni()}:{env_addr()[1]}"
        path = f"/.well-known/masque/udp/{host}/{port}/"
        headers = [
            (b":method", b"CONNECT"),
            (b":protocol", b"connect-udp"),
            (b":scheme", b"https"),
            (b":authority", authority.encode()),
            (b":path", path.encode()),
        ]
        credentials = basic_credentials()
        if with_credentials and credentials:
            headers.append((b"proxy-authorization", credentials))
        stream_id, fields = await self._request(headers)

        # RFC 9297 §3.2 and §3.4: the response advertises the capsule protocol
        # and must not frame a body length.
        assert fields.get(b"capsule-protocol") == b"?1", fields
        for name in (b"content-length", b"content-type", b"transfer-encoding"):
            assert name not in fields, f"{name} must be absent, got {fields}"
        return stream_id

    def send_stream(self, stream_id, data, end_stream=False):
        self._h3.send_data(stream_id, data, end_stream=end_stream)
        self.transmit()

    async def recv_stream(self, stream_id):
        """Waits for the next DATA chunk; returns (bytes, stream_ended)."""
        event = await self._next_event(stream_id)
        assert isinstance(event, DataReceived), f"expected data, got {event}"
        return event.data, event.stream_ended

    def send_udp(self, stream_id, payload, context_id=0):
        """Sends one HTTP Datagram with an explicit Context ID prefix.

        The prefix is applied by hand on purpose: RFC 9298 §5's "drop
        silently" rule for unknown Context IDs can only be tested by a client
        allowed to send one.
        """
        self._h3.send_datagram(stream_id, encode_uint_var(context_id) + payload)
        self.transmit()

    async def recv_udp(self, stream_id):
        """Waits for the next datagram on a session; returns its UDP payload."""
        event = await self._next_event(stream_id)
        assert isinstance(event, DatagramReceived), f"expected datagram, got {event}"
        buffer = Buffer(data=event.data)
        context_id = buffer.pull_uint_var()
        assert context_id == 0, f"context id {context_id} on a proxied payload"
        return event.data[buffer.tell() :]


def proxy_connection():
    """An async context manager holding one connection to the server under test."""
    host, port = env_addr()
    cert = os.environ.get("VOLTO_CERT")
    if not cert:
        sys.exit("VOLTO_CERT is not set: the certificate has to be trusted explicitly")

    configuration = QuicConfiguration(
        is_client=True,
        alpn_protocols=H3_ALPN,
        server_name=env_sni(),
        # QUIC datagram support is a transport parameter; without it the
        # DATAGRAM frames the UDP tests depend on cannot flow at all.
        max_datagram_frame_size=65536,
    )
    configuration.load_verify_locations(cafile=cert)
    return connect(host, port, configuration=configuration, create_protocol=ProxyClient)


async def start_tagged_udp_echo(tag):
    """A UDP target echoing each packet back with tag prepended.

    The tag turns "a packet came back" into "the packet came back from the
    target this session was opened for" — the Quarter-Stream-ID regression
    shape, same as the Go suite.
    """
    loop = asyncio.get_running_loop()

    class Echo(asyncio.DatagramProtocol):
        def connection_made(self, transport):
            self.transport = transport

        def datagram_received(self, data, addr):
            self.transport.sendto(bytes([tag]) + data, addr)

    transport, _ = await loop.create_datagram_endpoint(
        Echo, local_addr=("127.0.0.1", 0)
    )
    host, port = transport.get_extra_info("sockname")[:2]
    return transport, host, port


async def start_tagged_tcp_echo(tag):
    """A TCP target echoing each read back with tag prepended, closing on EOF."""

    async def serve(reader, writer):
        # A tunnel still open when its QUIC connection closes is aborted
        # towards the target, so a reset here is an ordinary ending.
        try:
            while True:
                data = await reader.read(65536)
                if not data:
                    break
                writer.write(bytes([tag]) + data)
                await writer.drain()
        except ConnectionError:
            pass
        finally:
            writer.close()

    server = await asyncio.start_server(serve, "127.0.0.1", 0)
    host, port = server.sockets[0].getsockname()[:2]
    return server, host, port


# --- tests -----------------------------------------------------------------


async def test_tcp_tunnel_round_trips():
    """The one path no other independent implementation exercises.

    Several round trips (a tunnel that is only right for the first write is a
    real failure mode), then the RFC 9114 §4.4 half-close: the client's FIN
    must reach the target as a write-shutdown, the target's close must come
    back as end of the response — and neither may tear the other down early.
    """
    server, host, port = await start_tagged_tcp_echo(0xD0)
    try:
        async with proxy_connection() as client:
            stream_id = await client.connect_tcp(f"{host}:{port}")

            for i in range(8):
                payload = f"volto-aioquic-tcp-{i}".encode()
                client.send_stream(stream_id, payload)
                got = b""
                want = bytes([0xD0]) + payload
                while len(got) < len(want):
                    data, ended = await client.recv_stream(stream_id)
                    assert not ended, "the tunnel ended mid-exchange"
                    got += data
                assert got == want, f"round {i}: {got!r} != {want!r}"

            # Half-close: FIN towards the target, EOF back from it.
            client.send_stream(stream_id, b"", end_stream=True)
            while True:
                data, ended = await client.recv_stream(stream_id)
                assert data == b"", f"unexpected data after FIN: {data!r}"
                if ended:
                    break
    finally:
        server.close()


async def test_tcp_tunnels_do_not_cross_talk():
    """TCP mirror of the Go suite's concurrent-sessions test.

    Two tunnels on one connection, both written before either is read, each
    reply checked for its own target's tag: a stream mix-up delivers to the
    wrong tunnel, not to nowhere.
    """
    tags = [0xD1, 0xD2]
    servers = []
    try:
        async with proxy_connection() as client:
            streams = []
            for tag in tags:
                server, host, port = await start_tagged_tcp_echo(tag)
                servers.append(server)
                streams.append(await client.connect_tcp(f"{host}:{port}"))

            for round_number in range(4):
                payloads = []
                for stream_id in streams:
                    payload = f"tunnel-{stream_id}-round-{round_number}".encode()
                    payloads.append(payload)
                    client.send_stream(stream_id, payload)

                for stream_id, payload, tag in zip(streams, payloads, tags):
                    got = b""
                    want = bytes([tag]) + payload
                    while len(got) < len(want):
                        data, ended = await client.recv_stream(stream_id)
                        assert not ended, "a tunnel ended mid-exchange"
                        got += data
                    assert got == want, f"{got!r} != {want!r} (misrouted tunnel data)"
    finally:
        for server in servers:
            server.close()


async def test_udp_session_round_trips():
    """Baseline: the tunnel carries UDP payloads both ways unmodified."""
    transport, host, port = await start_tagged_udp_echo(0xA0)
    try:
        async with proxy_connection() as client:
            stream_id = await client.connect_udp(host, port)
            for i in range(8):
                payload = f"volto-aioquic-udp-{i}".encode()
                client.send_udp(stream_id, payload)
                reply = await client.recv_udp(stream_id)
                want = bytes([0xA0]) + payload
                assert reply == want, f"round {i}: {reply!r} != {want!r}"
    finally:
        transport.close()


async def test_udp_sessions_do_not_cross_talk():
    """Three sessions on one connection: the Quarter-Stream-ID baseline."""
    tags = [0xB1, 0xB2, 0xB3]
    transports = []
    try:
        async with proxy_connection() as client:
            streams = []
            for tag in tags:
                transport, host, port = await start_tagged_udp_echo(tag)
                transports.append(transport)
                streams.append(await client.connect_udp(host, port))

            for round_number in range(4):
                payloads = []
                for stream_id in streams:
                    payload = f"session-{stream_id}-round-{round_number}".encode()
                    payloads.append(payload)
                    client.send_udp(stream_id, payload)

                for stream_id, payload, tag in zip(streams, payloads, tags):
                    reply = await client.recv_udp(stream_id)
                    want = bytes([tag]) + payload
                    assert reply == want, (
                        f"{reply!r} != {want!r} (a reply tagged for another "
                        "session means datagrams are misrouted)"
                    )
    finally:
        for transport in transports:
            transport.close()


async def test_unknown_context_id_is_dropped_silently():
    """RFC 9298 §5: an unknown Context ID is dropped, never answered and
    never fatal — the session must keep working afterwards."""
    transport, host, port = await start_tagged_udp_echo(0xC5)
    try:
        async with proxy_connection() as client:
            stream_id = await client.connect_udp(host, port)

            client.send_udp(stream_id, b"should-never-reach-the-target", context_id=7)
            client.send_udp(stream_id, b"should-arrive")

            reply = await client.recv_udp(stream_id)
            want = bytes([0xC5]) + b"should-arrive"
            assert reply == want, (
                f"{reply!r} != {want!r} (the context-7 datagram must be "
                "dropped, not forwarded or answered)"
            )
    finally:
        transport.close()


async def test_missing_credentials_are_challenged():
    """Credentials are checked before routing; omitting them draws a 407."""
    if not os.environ.get("VOLTO_USER"):
        print("  skipped: VOLTO_USER is not set, authentication is disabled")
        return
    transport, host, port = await start_tagged_udp_echo(0xC0)
    try:
        async with proxy_connection() as client:
            try:
                await client.connect_udp(host, port, with_credentials=False)
            except Refused as refusal:
                assert refusal.status == 407, f"status {refusal.status}, want 407"
                challenge = refusal.headers.get(b"proxy-authenticate")
                assert challenge == b'Basic realm="masque"', challenge
            else:
                raise AssertionError("an unauthenticated CONNECT-UDP succeeded")
    finally:
        transport.close()


async def test_denied_target_is_refused_with_proxy_status():
    """A policy refusal is a well-formed 403 with a machine-readable reason
    (RFC 9209), not a dropped stream."""
    async with proxy_connection() as client:
        try:
            await client.connect_udp("127.0.0.1", DENIED_PORT)
        except Refused as refusal:
            assert refusal.status == 403, f"status {refusal.status}, want 403"
            proxy_status = refusal.headers.get(b"proxy-status")
            assert proxy_status == b"volto; error=http_request_denied", proxy_status
        else:
            raise AssertionError(f"CONNECT-UDP to port {DENIED_PORT} succeeded")


TESTS = [
    test_tcp_tunnel_round_trips,
    test_tcp_tunnels_do_not_cross_talk,
    test_udp_session_round_trips,
    test_udp_sessions_do_not_cross_talk,
    test_unknown_context_id_is_dropped_silently,
    test_missing_credentials_are_challenged,
    test_denied_target_is_refused_with_proxy_status,
]


async def main():
    failures = 0
    for test in TESTS:
        print(f"=== {test.__name__}")
        try:
            await asyncio.wait_for(test(), timeout=60)
        except Exception:  # noqa: BLE001 - a test failure, whatever its type
            import traceback

            traceback.print_exc()
            failures += 1
            print(f"--- FAIL {test.__name__}")
        else:
            print(f"--- PASS {test.__name__}")
    if failures:
        sys.exit(f"{failures} of {len(TESTS)} tests failed")


if __name__ == "__main__":
    asyncio.run(main())
